### ISSUE TITLE
scylla doesn't exit on SIGTERM

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#909]
When running a jepsen's :slow-network test bed some of the nodes refuse to shutdown at the end of the test (stopped with `killall scylla` command). 

scylla HEAD: 7dc843fc1c8e4ea3fd8bcc3091f7397eb4e68837

**Comment by avikivity** [comment]
Please provide more information. Is there anything in the logs?  What is the system doing?

"it doesn't work" bug reports are not helpful.

**Comment by vladzcloudius** [comment]
On 02/15/2016 01:52 PM, Avi Kivity wrote:
>
> Please provide more information. Is there anything in the logs?
>

This happens when the log level is set to /error/. There is nothing in 
the logs then.
When I open all logs I hit the issue #768.

I'm going to fix the above first and hopefully the issue will reproduce 
with a /trace/ log level.

**Comment by nyh** [comment]
The "info" log level prints plenty of information on things which are shutting down - I don't think you need to go to "trace".

One wild guess on what might be happening in your case: something (repair, streaming, etc.?) from the hung node might have tried RPC to another node. But the other node was killed, and now when this node is shutting down, it waits for the RPC to be retransmitted again and again, which will never succeed.
I saw such a thing once when trying to kill many nodes at once while a repair was in progress. I should probably open a separate bug on that issue.

**Comment by vladzcloudius** [comment]
@nyh Nice guess. I rerun this with only _info_ log level and it indeed looks like it's stuck on RPC shutdown:
```
INFO  [shard 0] gossip - InetAddress 192.168.122.12 is now DOWN
INFO  [shard 0] repair - Starting shutdown of repair
INFO  [shard 0] repair - Completed shutdown of repair
INFO  [shard 0] storage_service - Drain on shutdown: starts
INFO  [shard 0] gossip - Announcing shutdown
INFO  [shard 0] storage_service - Node 192.168.122.13 state jump to normal
INFO  [shard 0] storage_service - Drain on shutdown: stop_gossiping done
```

**Comment by gleb-cloudius** [comment]
What do you mean by "RPC to be retransmitted again and again"? RPC does not retransmit anything by itself. If you try to send something to dead node RPC will try to establish a connection and until connection is established message will not be send. Shutting down rpc client show propagate to waiters given that our shutdown works for connecting socket. Avi, I remameber you've said our shutdown suppose to work on connecting socket, right?

**Comment by avikivity** [comment]
We don't have unconnected_socket yet, but we soon will.

**Comment by gleb-cloudius** [comment]
On Mon, Feb 15, 2016 at 06:32:48AM -0800, Avi Kivity wrote:
> We don't have unconnected_socket yet, but we soon will.
> 
If shutdown() works on connecting socket why do we need
unconnected_socket socket?
 
--
			Gleb.

**Comment by avikivity** [comment]
::connect() returns a future<connected_socket>, so there is no way to cause it to return an exception.

When we have

    struct unconnected_socket {
        future<connected_socket> connect();
        void abort();
    };

then abort() can call shutdown(2) to cancel the connection attempt.

**Comment by gleb-cloudius** [comment]
On Mon, Feb 15, 2016 at 06:50:29AM -0800, Avi Kivity wrote:
> ::connect() returns a future<connected_socket>, so there is no way to cause it to return an exception.
> 
> When we have
> 
>     struct unconnected_socket {
>         future<connected_socket> connect();
>         void abort();
>     };
> 
> then abort() can call shutdown(2) to cancel the connection attempt.
> 
Ah, right. Forgot about this our's API wart. Anyway according to the log in
this issue we stuck in cql server shutdown, not rpc.

--
			Gleb.

**Comment by nyh** [comment]
On Mon, Feb 15, 2016 at 3:12 PM, vladzcloudius <notifications@github.com>
wrote:

> @nyh <https://github.com/nyh> Nice guess. I rerun this with only *info*
> log level and it indeed looks like it's stuck on RPC shutdown:
>
> INFO  [shard 0] gossip - InetAddress 192.168.122.12 is now DOWN
> INFO  [shard 0] repair - Starting shutdown of repair
> INFO  [shard 0] repair - Completed shutdown of repair
> INFO  [shard 0] storage_service - Drain on shutdown: starts
> INFO  [shard 0] gossip - Announcing shutdown
> INFO  [shard 0] storage_service - Node 192.168.122.13 state jump to normal
> INFO  [shard 0] storage_service - Drain on shutdown: stop_gossiping done
>
>
If you look at storage_service::drain_on_shutdown(), you'll see that what
it does after this "stop_gossiping done" is call

     ss.shutdown_client_servers().get();

Because we didn't see a message that *that* is done, then definitely this
is the call that got stuck.

**Comment by nyh** [comment]
On Mon, Feb 15, 2016 at 4:27 PM, Gleb Natapov <notifications@github.com>
wrote:

> What do you mean by "RPC to be retransmitted again and again"? RPC does
> not retransmit anything by itself.
>
This probably has nothing to do with Vlad's original problem, but I'll
answer anyway - the "retransmission" I noticed was probably from
retransmission logic in the streaming code, not in the RPC code itself. So
I mis-spoke when "blaming" the RPC.

**Comment by slivne** [comment]
I think @asias sent a patch that was committed that fixes streaming behavior on shutdown - is this still open ?

**Comment by asias** [comment]
The fix:

commit 46c392eb17c1a5b4d13006bedf43fca477579b7c
Author: Asias He <asias@scylladb.com>
Date:   Tue Feb 2 21:01:22 2016 +0800

    messaging_service: Stop retrying if messaging_service is being shutdown

is included in what Vlad tested (scylla HEAD: 7dc843f).

**Comment by asias** [comment]
@nyh yes, it is stuck at shutdown of cql server. With pawel's fix I do not see the hang any more. The fix is also included in Vlad's test (scylla HEAD: 7dc843f). Perhaps we have more bugs in shutdown cql server. 

commit c90ec731c8c9ad0a9e0fd3b978f31641b3c3d69a
Author: Paweł Dziepak <pdziepak@scylladb.com>
Date:   Thu Feb 4 15:33:10 2016 +0100

    transport: do not close gate at connection shutdown

    connection::_pending_requests_gate is responsible for keeping connection
    objects alive as long as there are outstanding requests and is closed
    in connection::proccess() when needed. Closing it in connection::shutdown()
    as well may cause the gate to be closed twice what is a bug.

    Fixes #690.

    Signed-off-by: Paweł Dziepak <pdziepak@scylladb.com>
    Message-Id: <1454596390-23239-1-git-send-email-pdziepak@scylladb.com>

**Comment by asias** [comment]
@vladzcloudius I saw "[PATCH scylla v1] cql_server: destroy connection instances in stop()". Does it fix the cql hang you saw?

**Comment by vladzcloudius** [comment]
On Feb 19, 2016 4:05 AM, "Asias He" <notifications@github.com> wrote:
>
> @vladzcloudius I saw "[PATCH scylla v1] cql_server: destroy connection
instances in stop()". Does it fix the cql hang you saw?

@asias Yes. I think a commit message clearly states this fact.

**Comment by asias** [comment]
On Feb 19, 2016 12:59 PM, "vladzcloudius" <notifications@github.com> wrote:
>
> On Feb 19, 2016 4:05 AM, "Asias He" <notifications@github.com> wrote:
> >
> > @vladzcloudius I saw "[PATCH scylla v1] cql_server: destroy connection
> instances in stop()". Does it fix the cql hang you saw?
>
> @asias Yes. I think a commit message clearly states this fact.

I have some comments on the patch.

**Comment by vladzcloudius** [comment]
On 02/19/2016 07:26 AM, Asias He wrote:
> On Feb 19, 2016 12:59 PM, "vladzcloudius" <notifications@github.com> 
> wrote:
> >
> > On Feb 19, 2016 4:05 AM, "Asias He" <notifications@github.com> wrote:
> > >
> > > @vladzcloudius I saw "[PATCH scylla v1] cql_server: destroy connection
> > instances in stop()". Does it fix the cql hang you saw?
> >
> > @asias Yes. I think a commit message clearly states this fact.
>
> I have some comments on the patch.

I saw them and I have replied to your comments.

**Comment by asias** [comment]
@vladzcloudius any updates on the newer version of your fix?

**Comment by vladzcloudius** [comment]
On 02/24/2016 09:06 AM, Asias He wrote:
>
> @vladzcloudius <https://github.com/vladzcloudius> any updates on the 
> newer version of your fix
>

@asias Still hunting a root cause. The /epoll_pwait()/ doesn't return 
for some connections (their underlying sockets) after they are supposed 
to be /shutdown()/). A sample application  I cooked proved that when a 
connection is shutdown() for read or for read-and-write the 
corresponding /epoll_wait()/ immediately returns with a EPOLLHUP event. 
However after the sockets in question are shutdown() no /epoll_pwait()/ 
returns an EPOLLHUP. I tried to run with /strace/ but the problem didn't 
reproduce with it (ever though I minimized the /strace/ output to only 
/shutdown/ and /epoll_pwait/ syscalls).

One thing that could explain this would be that we /close()/ the socket 
before /shutdown(). /In this case the socket is immediately removed from 
the poll() list, however since we never /close()/ the posix sockets this 
possibility is ruled out. I'll try to check is the corresponding sockets 
are not somehow removed from the poll() list. /strace /would 
significantly simplify the debugging but unfortunately it changes the 
timing the way the problem doesn't reproduce...

**Comment by vladzcloudius** [comment]
We've found a root cause. There are a few issues actually:
1. storage_service::token_metadata and gossiper::endpoint_state_map are not updated as an atomic unit when they are shadowed to non-zero shards. As a result there was a state when a bootstrapping node was already in a token_metadata::_bootstrap_tokens but not yet in an gossiper::endpoint_state_map. 

This is not an impossible since this may happen when a bootstrapping node has crashed before it complete the bootstrapping. However this should not happen in this scenario since a node is bootstrapping just fine and having a few requests failed just because there is a bootstrapping node is not an expected user experience.

2. Write requests related code in a storage_proxy was not handling the case when a `pending` node is also `dead` correctly. It was not sending any request to such a node but was still waiting for a response count to hit the mark as if there should arrive a response for such a node as well. As a result such write request was never becoming complete nor timing out. 

A side effect of this is an issue this ticket was originally opened on: when a node is being killed the cql_server::connection::_pending_requests_gate of the corresponding CQL connection was never dropping to zero and as a result we were waiting for it forever ignoring further signals.

I've opened an additional issue to track this issue specifically - issue #965. 
@gleb-cloudius FYI

**Comment by asias** [comment]
@vladzcloudius so with the patch to sync endpoint_data_map and token_metadata + f59415b3c6ebdd8f85b72f5658767ac676493845 , you do not see hang of cql server anymore?

@gleb-cloudius Is #965 fixed? I see it is still open. I think your patch already merged (commit f59415b3c6ebdd8f85b72f5658767ac676493845) which fixes #965.

**Comment by slivne** [comment]
Guys if there are patches not on head lets merge them - we can not have
items closed for patches that are not merged to head

**Comment by gleb-cloudius** [comment]
On Tue, Mar 08, 2016 at 11:37:37PM -0800, Asias He wrote:
> @vladzcloudius so with the patch to sync endpoint_data_map and token_metadata + f59415b3c6ebdd8f85b72f5658767ac676493845 , you do not see hang of cql server anymore?
> 
> @gleb-cloudius Is #965 fixed? I see it is still open. I think your patch already merged (commit f59415b3c6ebdd8f85b72f5658767ac676493845) which fixes #965.
> 
Shlomi reopened it because test still failed. The problem is in the
test, so after test will be fixed bug will be closed.

--
			Gleb.

**Comment by vladzcloudius** [comment]
On 03/09/2016 09:37 AM, Asias He wrote:
>
> @vladzcloudius <https://github.com/vladzcloudius> so with the patch to 
> sync endpoint_data_map and token_metadata + f59415b 
> <https://github.com/scylladb/scylla/commit/f59415b3c6ebdd8f85b72f5658767ac676493845> 
> , you do not see hang of cql server anymore?
>

@asias - Of course. Why would I mark the above commit "Fixes #909" if it 
didn't? ;)
